### ISSUE TITLE
ttnn.max_bw / ttnn.min_bw return zero gradients for equal infinities

### ISSUE THREAD

**Issue by kanapitsas** [tenstorrent/tt-metal#54562]
### Summary

`ttnn.max_bw` and `ttnn.min_bw` return zero gradients when both inputs are the same
infinity.

With `grad = 1`:

| input | other | PyTorch golden | current expression |
|---|---|---|---|
| `+inf` | `+inf` | `[0.5, 0.5]` | `[0, 0]` |
| `-inf` | `-inf` | `[0.5, 0.5]` | `[0, 0]` |
| `2.0` | `2.0` | `[0.5, 0.5]` | `[0.5, 0.5]` |

Both operations use the same `_min_or_max_bw` implementation.

### Root cause

The code compares the inputs through a subtraction:

```cpp
Tensor t_sub = ttnn::subtract(other, input, ...);
Tensor t_sub_gtz = ttnn::gtz(t_sub, ...);
Tensor t_sub_eqz = ttnn::eqz(t_sub, ...);
Tensor t_sub_ltz = ttnn::ltz(t_sub, ...);
```

For equal infinities, `other - input` is `NaN`. All three predicates return zero for
`NaN`, so both gradient masks are zero, and the gradient is not routed to a default
branch — a sum of masked products has none.

The current code is in
[`binary_backward.cpp`](https://github.com/tenstorrent/tt-metal/blob/1ec5b72398130fab0331d47862cd943131570907/ttnn/cpp/ttnn/operations/eltwise/binary_backward/binary_backward.cpp#L56-L70).

### Why the input is legitimate

`-inf` is not an accident in this position: it is the standard masked-attention fill value,
the result of `log(0)`, and the score assigned to a forbidden action. Two masked positions
meeting under an elementwise `maximum`/`minimum` is the ordinary way two equal infinities
arrive at this op — and the gradient is then lost silently, with no `NaN` or `Inf` in the
output to notice.

### Model of the current expression

This script does not call `ttnn`. It evaluates the same mask arithmetic in Python and
compares it with the registered golden, so the discrepancy can be checked on a host with no
Tenstorrent device.

```python
import torch


def current(op, x, y, grad=1.0):
    delta = y - x
    gt = float(delta > 0.0)
    eq = float(delta == 0.0)
    lt = float(delta < 0.0)
    first = lt * grad + eq * grad / 2.0
    second = gt * grad + eq * grad / 2.0
    return [first, second] if op == "max" else [second, first]


def expected(op, x, y, grad=1.0):
    a = torch.tensor([x], dtype=torch.float32, requires_grad=True)
    b = torch.tensor([y], dtype=torch.float32, requires_grad=True)
    getattr(torch, op)(a, b).backward(torch.tensor([grad]))
    return [a.grad.item(), b.grad.item()]


for op in ("max", "min"):
    for value in (float("inf"), float("-inf"), 2.0):
        print(op, value, "current:", current(op, value, value),
              "expected:", expected(op, value, value))
```

Output:

```text
max inf current: [0.0, 0.0] expected: [0.5, 0.5]
max -inf current: [0.0, 0.0] expected: [0.5, 0.5]
max 2.0 current: [0.5, 0.5] expected: [0.5, 0.5]
min inf current: [0.0, 0.0] expected: [0.5, 0.5]
min -inf current: [0.0, 0.0] expected: [0.5, 0.5]
min 2.0 current: [0.5, 0.5] expected: [0.5, 0.5]
```

### Fix

Compare `other` and `input` directly and drop `t_sub`:

```cpp
Tensor t_gt = ttnn::gt(other, input, std::nullopt, output_mem_config);
Tensor t_eq = ttnn::eq(other, input, std::nullopt, output_mem_config);
Tensor t_lt = ttnn::lt(other, input, std::nullopt, output_mem_config);
```

This fixes both ops, since they share `_min_or_max_bw`, and it removes a dispatch rather
than adding one.

**The binary comparisons are genuine comparison sequences here, not a subtract followed by
a zero-test.** `is_binary_sfpu_op` routes `EQ`/`GT`/`LT` to the SFPU path whenever both
operands share a dtype in `{FLOAT32, BFLOAT16, INT32, UINT16, UINT32}`
(`binary_ng_device_operation.cpp` L69-L74), and `OpConfig` then emits `eq_binary_tile`,
`gt_binary_tile` and `lt_binary_tile` (`binary_ng_utils.cpp` L511-L545). That set covers
exactly the dtypes in which `±inf` is representable. `calculate_binary_comp_fp32_equal`
rejects a `NaN` operand through `SFPIADD(inf, |a| + |b|, CC_GTE0)`; two equal infinities are
not rejected, because `|inf| + |inf| = inf` is not greater than `inf`, so the comparison
runs and reports equality — your own semantics table records this and marks it green on
Wormhole (`tt_metal/tt-llk/tests/python_tests/helpers/sfpu_domains.py`, `SfpuElwEq` and
`SfpuElwLt`). #53908's measured table shows the same from the other side: `ttnn.eq(x, inf)`
returns 1 at `x = inf`.

For a dtype outside that set, `EQ` falls back to `postprocess = EQZ` over the subtraction
(`binary_ng_utils.cpp` L179-L213) — the same construction as today, so the change is neutral
there rather than a regression.

**NaN routing is unchanged, deliberately.** If either operand is `NaN`, all three
comparisons return `0` — the same three zeros the current code produces through the rejected
subtraction — so both gradients stay `0`, exactly as today. This keeps the change clear of
#51470, which is about `NaN` in the forward ops and a different mechanism.

Please also add exact-tie tests for finite values, `+inf` and `-inf`. The current tests use
disjoint random ranges (`test_backward_max.py` draws `input` from `[-100, -1]` and `other`
from `[1, 2]`), so a tie can never occur.

### Validation limits

I reproduced the two device primitives on `ttsim` for Wormhole and Blackhole. I have not
run `max_bw` or `min_bw` end-to-end on Tenstorrent hardware.

**Comment by kanapitsas** [comment]
Both ops share _min_or_max_bw, so a single change fixes them. I'd like to prepare the patch and the exact-tie tests. Could you assign this issue to me before I open the PR?
